File: src/main/java/com/chessonline/MoveHandler.java
package com.chessonline;

public class MoveHandler {
    private ChessBoard chessBoard;

    public MoveHandler(ChessBoard chessBoard) {
        this.chessBoard = chessBoard;
    }

    public boolean movePiece(String fromPosition, String toPosition) {
        int[] fromCoords = chessBoard.positionToCoordinates(fromPosition);
        int[] toCoords = chessBoard.positionToCoordinates(toPosition);
    
        if (chessBoard.isDestinationEmpty(toPosition)) {
            return handleMove(fromCoords, toCoords, toPosition);
        }
        
        chessBoard.isCheckmate(chessBoard.getWhiteKing());
        
        return handleCapture(toCoords, fromCoords, toPosition);
    }

    private boolean handleMove(int[] fromCoords, int[] toCoords, String toPosition) {
        Piece piece = chessBoard.getBoard()[fromCoords[0]][fromCoords[1]];

        if (!piece.move(toPosition)) {
            return false;
        }
    
        chessBoard.updateBoard(fromCoords, toCoords, piece);

        if (piece instanceof Pawn) {
            int lastRow = piece.getColor().equals("white") ? 8 : 1;
            if (toPosition.charAt(1) - '0' == lastRow) {
                chessBoard.promotePawn((Pawn) piece);
            }
        }

        chessBoard.isCheckmate(chessBoard.getWhiteKing());

        return true;
    }

    private boolean handleCapture(int[] toCoords, int[] fromCoords, String toPosition) {
        Piece piece = chessBoard.getBoard()[fromCoords[0]][fromCoords[1]];
        Piece targetPiece = chessBoard.getBoard()[toCoords[0]][toCoords[1]];
        
        if (piece.getColor().equals(targetPiece.getColor())) return false;

        if (!piece.isCaptureMovementValid(toPosition)) return false;

        chessBoard.removePiece(targetPiece);
        chessBoard.updateBoard(fromCoords, toCoords, piece);

        return true;
    }
}

File: engine/helpers/move_handler.py
from ..pieces.pawn import Pawn


class MoveHandler:
    def __init__(self, chess_board):
        self.chess_board = chess_board

    def move_piece(self, from_position, to_position):
        from_coords = self.chess_board.position_to_coordinates(from_position)
        to_coords = self.chess_board.position_to_coordinates(to_position)

        if self.chess_board.is_destination_empty(to_position):
            return self.handle_move(from_coords, to_coords, to_position)

        # checkmate check (same as Java, even if odd placement)
        self.chess_board.is_checkmate(self.chess_board.get_white_king())

        return self.handle_capture(to_coords, from_coords, to_position)

    def handle_move(self, from_coords, to_coords, to_position):
        piece = self.chess_board.get_board()[from_coords[0]][from_coords[1]]

        if not piece.move(to_position):
            return False

        self.chess_board.update_board(from_coords, to_coords, piece)

        # pawn promotion
        if isinstance(piece, Pawn):
            last_row = 8 if piece.get_color() == "white" else 1
            if int(to_position[1]) == last_row:
                self.chess_board.promote_pawn(piece)

        self.chess_board.is_checkmate(self.chess_board.get_white_king())

        return True

    def handle_capture(self, to_coords, from_coords, to_position):
        piece = self.chess_board.get_board()[from_coords[0]][from_coords[1]]
        target_piece = self.chess_board.get_board()[to_coords[0]][to_coords[1]]

        if piece.get_color() == target_piece.get_color():
            return False

        if not piece.is_capture_movement_valid(to_position):
            return False

        self.chess_board.remove_piece(target_piece)
        self.chess_board.update_board(from_coords, to_coords, piece)

        return True